pkg/wal/wal.go

```diff
@@ -0,0 +1,345 @@
+		mutableStore: mutableStore,
+		walStore: walStore,
+		tokenGeneratorPath: tokenGeneratorPath,
+		maxConcurrency: maxConcurrency,
+		l: logger,
+	}
+	for _, option := range options {
+		option(&wal)
+	}
+	// Check if token generator object exists
+	_ = mutableStore.Put(context.Background(), wal.tokenGeneratorPath, strings.NewReader(""), storage.OverWrite)
+	return &wal
+}
+
+// Gets a token such that tokens are K-sortable.
+func (w *WAL) getToken(ctx context.Context) (string, error) {
+	err := w.updateTokenTimestamp(ctx)
+	if err != nil {
+		return "", fmt.Errorf("failed to update the token generator err:%s", err.Error())
+	}
+
+	// Read the updateTime from the tokenGenerator object
+	attr, err := w.mutableStore.GetAttr(ctx, w.tokenGeneratorPath)
+	if err != nil {
+		return "", fmt.Errorf("failed to get tokenGenerator attributes: %s, err:%s", w.tokenGeneratorPath, err)
+	}
+
+	// Use the updateTime from the token generator (gain independence from local wall clock)
+	k, err := ksuid.NewRandomWithTime(attr.Updated)
+	if err != nil {
+		return "", fmt.Errorf("failed tp generate ksuid: %s", err)
+	}
+
+	w.l.Debug("generated token", zap.String("token", k.String()), zap.Time("updateTime", attr.Updated))
+
+	return k.String(), nil
+}
+
+// Adds a WAL entry to WAL
+func (w *WAL) Add(ctx context.Context, p string) error {
+	e := Entry{
+		Payload: p,
+	}
+	var err error
+	e.Token, err = w.getToken(ctx)
+	if err != nil {
+		w.l.Error("failed to generate token for entry", zap.Error(err), zap.String("payload", p))
+		return fmt.Errorf("failed to generate token: %v", err)
+	}
+	err = w.walStore.Put(ctx, e.Token, strings.NewReader(e.Payload), storage.NoOverWrite) // Should be a new entry
+	if err != nil {
+		w.l.Error("failed to add wal entry", zap.Error(err), zap.String("token", e.Token))
+		return fmt.Errorf("failed to add wal token: %s, entry: %s", e.Token, err.Error())
+	}
+	w.l.Info("Write wal entry", zap.String("token", e.Token))
+	return err
+}
+
+func (w *WAL) GetExpirationDuration() time.Duration {
+	return 10 * time.Minute
+}
+
+func (w *WAL) updateTokenTimestamp(ctx context.Context) error {
+	return w.mutableStore.Touch(ctx, w.tokenGeneratorPath)
+}
+
+// Reads the WAL starting from the tokens passed in. If startFrom is empty it will entry from beginning.
+// Repeated reads can include duplicate tokens. No tokens are missed
+// Returns false if the list has more entries that can be listed.
+// Use the last Entry of the previous call to paginate to the next set of keys.
+func (w *WAL) ListTokens(ctx context.Context, fromToken string, max int) (tokens []string, next string, err error) {
+	k, err := ksuid.Parse(fromToken)
+	if err != nil {
+		return nil, "", err
+	}
+	if max > 1000 {
+		max = 1000
+	}
+
+	// Go back in time to include the keys that might have been written before the token was generated.
+	ksuidOld := k.Time().Add(-w.GetExpirationDuration() * 2)
+	tokens, next, err = w.walStore.KeysPrefix(ctx, ksuidOld.String(), "", "", max)
```

Comment:
i'm slightly confused... `ksuidOld` is a `Time` instance.

b/c both the `Time` here and the `KSUID` in `getToken` that's used in the corresponding`walStore.Put` in `Add` are used _via casts to string_, some info that might otherwise be decipherable via the type system is probably being lost on me.

at the possible expense of DSLification, could there be some functions added to cast ksuids and times to strings?  or perhaps some comments on the relation between the time string and the ksuid string would suffice?

Reply:
I should port the design over from the internal documents here.
The main idea is to give a model for the clients of WAL to write their intent to the WAL, get an expiration limit in which they have to finish the work. This is to allow for a listing of WAL that does not need to start from the first entry in the WAL (clients writing to back in time in the WAL due to asynchronous model in play).

Reply:
since i have seen [the format of random ksuids with time](https://github.com/segmentio/ksuid/blob/master/ksuid.go#L217), namely `<time><rand>`, i now understand the cohesion<>decoupling tradeoffs involved in the string casts and consider it maintainable.

still some slight vexation about what amounts to explicitly phrasing assumptions about the upstream deps within the datamon source, yet any such are far from blocking.